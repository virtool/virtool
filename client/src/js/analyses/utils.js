import {
    compact,
    fill,
    flatMap,
    fromPairs,
    keys,
    map,
    max,
    maxBy,
    min,
    minBy,
    reject,
    sortBy,
    sumBy,
    toNumber,
    uniq
} from "lodash-es";
import { formatIsolateName } from "../utils/utils";

const calculateAnnotatedOrfCount = orfs => reject(orfs, { "hits.length": 0 }).length;

const calculateORFMinimumE = hits => {
    if (hits.length === 0) {
        return;
    }

    const minHit = minBy(hits, "full_e");
    return minHit.full_e;
};

const calculateSequenceMinimumE = orfs => {
    if (orfs.length === 0) {
        return;
    }

    const minEValues = map(orfs, orf => calculateORFMinimumE(orf.hits));
    return min(minEValues);
};

export const extractFamilies = orfs => {
    const families = uniq(flatMap(orfs, orf => flatMap(orf.hits, hit => keys(hit.families))));
    return reject(families, f => f === "None");
};

export const extractNames = orfs => {
    return uniq(flatMap(orfs, orf => flatMap(orf.hits, hit => hit.names)));
};

/**
 * Transform an array of coordinate pairs into an flat array where the index is the x coordinate and the value is the y
 * coordinate.
 *
 * @param {Array} align - the coordinates
 * @param {Number} length - the length of the generated flat array
 * @returns {Array} - the flat array
 *
 */
export const fillAlign = ({ align, length }) => {
    const filled = Array(length);

    if (!align) {
        return fill(filled, 0);
    }

    const coords = fromPairs(align);

    let prev = 0;

    return map(filled, (depth, i) => {
        if (coords.hasOwnProperty(i)) {
            prev = coords[i];
        }

        return prev;
    });
};

export const formatData = detail => {
    if (detail.workflow === "pathoscope_bowtie") {
        return formatPathoscopeData(detail);
    }

    if (detail.workflow === "nuvs") {
        return formatNuVsData(detail);
    }

    if (detail.workflow === "aodp") {
        return formatAODPData(detail);
    }
};

export const formatAODPData = detail => {
    const results = map(detail.results, result => {
        const isolates = map(result.isolates, isolate => {
            const sequences = map(isolate.sequences, sequence => {
                return {
                    ...sequence,
                    identities: getSequenceIdentities(sequence)
                };
            });

            return {
                ...isolate,
                sequences,
                identities: getIdentities(sequences)
            };
        });

        const identities = getIdentities(isolates);

        return {
            ...result,
            isolates,
            identities,
            identity: max(identities)
        };
    });

    return { ...detail, results };
};

const getIdentities = data => flatMap(data, item => item.identities);

const getSequenceIdentities = sequence => flatMap(sequence.hits, hit => hit.identity);

export const formatNuVsData = detail => {
    const results = map(detail.results, result => ({
        ...result,
        id: toNumber(result.index),
        annotatedOrfCount: calculateAnnotatedOrfCount(result.orfs),
        e: calculateSequenceMinimumE(result.orfs),
        families: extractFamilies(result.orfs),
        names: extractNames(result.orfs)
    }));

    const longestSequence = maxBy(results, result => result.sequence.length);

    const { cache, created_at, id, ready, user, workflow } = detail;

    return {
        cache,
        created_at,
        id,
        ready,
        results,
        user,
        workflow,
        maxSequenceLength: longestSequence.sequence.length
    };
};

export const formatSequence = (sequence, readCount) => ({
    ...sequence,
    filled: fillAlign(sequence),
    reads: sequence.pi * readCount
});

export const formatPathoscopeData = detail => {
    if (detail.results.length === 0) {
        return detail;
    }

    const {
        cache,
        created_at,
        results,
        id,
        index,
        read_count,
        ready,
        reference,
        subtractions,
        user,
        workflow
    } = detail;

    const formatted = map(results, otu => {
        const isolateNames = [];

        // Go through each isolate associated with the OTU, adding properties for weight, read count,
        // median depth, and coverage. These values will be calculated from the sequences owned by each isolate.
        const isolates = map(otu.isolates, isolate => {
            // Make a name for the isolate by joining the source type and name, eg. "Isolate" + "Q47".
            const name = formatIsolateName(isolate);

            isolateNames.push(name);

            const sequences = sortBy(
                map(isolate.sequences, sequence => formatSequence(sequence, read_count)),
                "length"
            );

            const filled = flatMap(sequences, "filled");

            // Coverage is the number of non-zero depth positions divided by the total number of positions.
            const coverage = compact(filled).length / filled.length;

            return {
                ...isolate,
                name,
                filled,
                coverage,
                sequences,
                maxDepth: max(filled),
                pi: sumBy(sequences, "pi"),
                depth: median(filled)
            };
        });

        const filled = mergeCoverage(isolates);
        const pi = sumBy(isolates, "pi");

        return {
            ...otu,
            filled,
            pi,
            isolates: sortBy(isolates, "coverage").reverse(),
            coverage: maxBy(isolates, "coverage").coverage,
            depth: median(filled),
            isolateNames: reject(uniq(isolateNames), "Unnamed Isolate"),
            maxGenomeLength: maxBy(isolates, "filled.length").length,
            maxDepth: maxBy(isolates, "maxDepth").maxDepth,
            reads: pi * read_count
        };
    });

    return {
        cache,
        created_at,
        id,
        index,
        reference,
        results: formatted,
        read_count,
        ready,
        subtractions,
        subtractedCount: detail.subtracted_count,
        user,
        workflow
    };
};

export const fuseSearchKeys = {
    pathoscope_bowtie: ["name", "abbreviation"],
    nuvs: ["families", "names"],
    aodp: ["name"]
};

/**
 * Calculate the median of an Array of numbers.
 *
 * @param values - an array of numbers
 * @returns {number|*} - the median
 */
export const median = values => {
    const sorted = values.slice().sort((a, b) => a - b);

    const midIndex = (sorted.length - 1) / 2;

    if (midIndex % 1 === 0) {
        return sorted[midIndex];
    }

    const lowerIndex = Math.floor(midIndex);
    const upperIndex = Math.ceil(midIndex);

    return Math.round((sorted[lowerIndex] + sorted[upperIndex]) / 2);
};

/**
 * Merge the coverage arrays for the given isolates. This is used to render a representative coverage chart for the
 * parent OTU.
 *
 * @param isolates
 * @returns {Array}
 */
export const mergeCoverage = isolates => {
    const longest = maxBy(isolates, isolate => isolate.filled.length);
    const coverages = map(isolates, isolate => isolate.filled);
    return map(longest.filled, (depth, index) => max(map(coverages, coverage => coverage[index])));
};
