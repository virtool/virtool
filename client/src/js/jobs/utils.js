import { get, startCase } from "lodash-es";

const importResultsDescription = {
    title: "Save Results",
    description: "Annotate the results and save them in the database."
};

const makeAnalysisDirDescription = {
    title: "Create Directory",
    description: "Create a directory where the job will run and analysis files will be stored."
};

const prepareReadsDescription = {
    title: "Prepare Reads",
    description: "Get existing trimmed data or create a cache."
};

const prepareQCDescription = {
    title: "Check QC",
    description: "Run FastQC if a new cache was created."
};

const aodpDescriptions = {
    make_analysis_dir: makeAnalysisDirDescription,
    prepare_reads: prepareReadsDescription,
    join_reads: {
        title: "Join Reads",
        description: "Join paired end reads into single long reads."
    },
    deduplicate_reads: {
        title: "Deduplicate Reads",
        description: "Remove duplicate reads and quantify duplication."
    },
    prepare_index: {
        title: "Prepare Index",
        description: "Prepare the reference index for searching."
    },
    aodp: {
        title: "AODP",
        description: "Run AODP using the amplicon and reference data."
    },
    import_results: importResultsDescription
};

const buildIndexDescriptions = {
    mk_index_dir: {
        title: "Create Directory",
        description: "Create data directory to store the new Bowtie2 index in."
    },
    write_fasta: {
        title: "Generate OTU FASTA",
        description: "Generate a FASTA file containing sequences for all default OTU isolates."
    },
    bowtie_build: {
        title: "Bowtie Build",
        description: "Build a Bowtie2 index from the FASTA file generated in the previous step."
    },
    replace_old: {
        title: "Replace Previous",
        description: "Make the new build the active index for the reference."
    }
};

const createSampleDescriptions = {
    make_sample_dir: {
        title: "Create Directory",
        description: "Create a data directory for the new sample."
    },
    copy_files: {
        title: "Copy Files",
        description: "Copy and compress input FASTQ files."
    },
    fastqc: {
        title: "FastQC",
        description: "Run FastQC on the sample FASTQ files."
    },
    parse_fastqc: {
        title: "Parse FastQC",
        description: "Parse FastQC output and store in database."
    },
    clean_watch: {
        title: "Cleanup",
        description: "Remove the source files."
    }
};

const createSubtractionDescriptions = {
    make_subtraction_dir: {
        title: "Create Directory",
        description: "Create a data directory for the new subtraction."
    },
    unpack: {
        title: "Unpack",
        description: "Decompress the sequence file if necessary."
    },
    set_stats: {
        title: "Save Sequence Statistics",
        description: "Save the GC content and sequence count to the database."
    },
    bowtie_build: {
        title: "Bowtie Build",
        description: "Build a Bowtie2 index from the FASTA file provided by the user."
    },
    compress: {
        title: "Compress FASTA",
        description: "Compress the FASTA data for long-term storage."
    }
};

const pathoscopeBowtieDescriptions = {
    make_analysis_dir: makeAnalysisDirDescription,
    prepare_reads: prepareReadsDescription,
    prepare_qc: prepareQCDescription,
    map_default_isolates: {
        title: "Map to Default Isolates",
        description: "Map the sequences of all OTU default isolates."
    },
    generate_isolate_fasta: {
        title: "Write Isolate FASTA",
        description: "Create a FASTA file containing all sequences of OTUs detected in previous step."
    },
    build_isolate_index: {
        title: "Build Isolate Index",
        description: "Build a Bowtie2 index from the all-isolate FASTA generated in the previous step."
    },
    map_isolates: {
        title: "Map to Isolates",
        description: "Map to the all-isolate index."
    },
    map_subtraction: {
        title: "Map to Subtraction",
        description: "Map all reads mapped in the previous step to the subtraction."
    },
    subtract_mapping: {
        title: "Subtract Reads",
        description: "Remove reads that have a better alignment against the subtraction than an OTU."
    },
    pathoscope: {
        title: "Pathoscope",
        description: "Reassign ambiguous mappings using Pathoscope 2.0."
    },
    import_results: importResultsDescription,
    cleanup_indexes: {
        title: "Clean-up Indexes",
        description: "Prune stale OTU indexes if they are no longer being used."
    }
};

const nuvsDescriptions = {
    make_analysis_dir: makeAnalysisDirDescription,
    prepare_reads: prepareReadsDescription,
    prepare_qc: prepareQCDescription,
    eliminate_otus: {
        title: "Eliminate OTUs",
        description: "Map against the OTU index and eliminate reads that match."
    },
    eliminate_subtraction: {
        title: "Eliminate Subtraction",
        description: "Map against the subtraction index and eliminate reads that match."
    },
    reunite_pairs: {
        title: "Reunite Pairs",
        description: "Ensure all remaining read pairs are paired correctly after subtraction."
    },
    assemble: {
        title: "Assemble",
        description: "Assemble contigs from remaining reads using SPAdes."
    },
    process_fasta: {
        title: "Process Assembly",
        description: "Find significant open reading frames (ORFs) in the assembled contigs."
    },
    prepare_hmm: {
        title: "Prepare HMM Profiles",
        description: "Run hmmpress on profiles in preparation for running hmmscan."
    },
    vfam: {
        title: "VFam",
        description: "Use hmmscan and vfam profiles to find viral motifs in the assembled contigs."
    },
    import_results: importResultsDescription
};

export const stepDescriptions = {
    aodp: aodpDescriptions,
    build_index: buildIndexDescriptions,
    create_sample: createSampleDescriptions,
    create_subtraction: createSubtractionDescriptions,
    pathoscope_bowtie: pathoscopeBowtieDescriptions,
    nuvs: nuvsDescriptions
};

export const getStepDescription = (stage, state, task) => {
    if (state === "waiting") {
        return {
            title: "Waiting",
            description: "Waiting for resources to become available."
        };
    }

    if (state === "complete") {
        return {
            title: "Complete",
            description: ""
        };
    }

    if (state === "cancelled") {
        return {
            title: "Cancelled",
            description: ""
        };
    }

    return get(stepDescriptions, [task, stage], {
        description: "",
        title: startCase(stage)
    });
};
