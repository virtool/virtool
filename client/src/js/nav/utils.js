import { includes } from "lodash-es";

export const excludePaths = (paths = []) => {
    return function (match, location) {
        if (includes(paths, location.pathname)) {
            return false;
        }

        return !!match;
    };
};
