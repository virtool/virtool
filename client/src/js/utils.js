import Numeral from "numeral";
import { sampleSize, get, startCase, capitalize } from "lodash";

export const taskDisplayNames = {
    nuvs: "NuVs",
    pathoscope_bowtie: "PathoscopeBowtie",
    pathoscope_snap: "PathoscopeSNAP"
};

export const getTaskDisplayName = (taskPrefix) => get(taskDisplayNames, taskPrefix, startCase(taskPrefix));

export const numberDictionary = {
    0: "zero",
    1: "one",
    2: "two",
    3: "three",
    4: "four",
    5: "five",
    6: "six",
    7: "seven",
    8: "eight",
    9: "nine"
};

const alphanumeric = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789";

export const createRandomString = (length=8) => {
    return sampleSize(alphanumeric, length).join("")
};

export const numberToWord = (number) => numberDictionary[Number(number)] || number;

export const byteSize = bytes => Numeral(bytes).format("0.0 b");

export const toScientificNotation = (number) => {
    if (number < 0.01 || number > 1000) {
        const split = number.toExponential().split("e");
        const exponent = split[1].replace("+", "");
        return Numeral(split[0]).format("0.00") + "â‚‘" + exponent;
    }

    return Numeral(number).format("0.000");
};

export const formatIsolateName = (isolate) => {
    if (
        isolate.source_type && isolate.source_type !== "unknown" ||
        isolate.sourceType && isolate.sourceType !== "unknown"
    ) {
        return (
            capitalize(isolate.source_type || isolate.sourceType) + " "
            + (isolate.source_name || isolate.sourceName)
        );
    }

    return "Unnamed";
};
