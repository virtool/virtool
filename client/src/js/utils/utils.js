/**
 * General utility constants and functions.
 *
 */
import { capitalize, get, replace, sampleSize, split, startCase, upperFirst } from "lodash-es";
import numbro from "numbro";
import { getAccountAdministrator } from "../account/selectors";

/**
 * A string containing all alphanumeric digits in both cases.
 *
 * @type {string}
 */
export const alphanumeric = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789";

/**
 * Converts an integer in bytes to a nicely formatted string (eg. 10.2 GB).
 *
 * @func
 * @param bytes {number}
 * @returns {string}
 */
export const byteSize = bytes => {
    if (bytes) {
        return numbro(bytes).format({ output: "byte", base: "decimal", mantissa: 1 });
    }

    return "0.0B";
};

/**
 * Create a URL object given a find term or a page number. Both parameters are optional.
 *
 * @func
 * @param term {string} a search string to place in the URL
 * @returns {URL}
 */
export const createFindURL = term => {
    const url = new window.URL(window.location);

    if (term !== undefined) {
        if (term) {
            url.searchParams.set("find", term);
        } else {
            url.searchParams.delete("find");
        }
    }

    return url;
};

/**
 * Create a random string of {@link length} from [alphanumeric]{@link module:utils.alphanumeric}.
 *
 * @func
 * @param length {number} the length of string to return
 */
export const createRandomString = (length = 8) => sampleSize(alphanumeric, length).join("");

/**
 * Download the file at the given {@link path}.
 *
 * @func
 * @param path {string}
 */
export const followDownload = path => {
    const a = document.createElement("A");
    a.href = path;
    a.download = path.substr(path.lastIndexOf("/") + 1);
    document.body.appendChild(a);
    a.click();
    document.body.removeChild(a);
};

/**
 * Download a file with the given ``filename`` with the given ``text`` content. This allows downloads of
 * dynamically generated files.
 *
 * @func
 * @param filename
 * @param text
 */
export const followDynamicDownload = (filename, text) => {
    const a = document.createElement("a");
    a.href = `data:text/plain;charset=utf-8,${encodeURIComponent(text)}`;
    a.download = filename;

    a.style.display = "none";
    document.body.appendChild(a);

    a.click();

    document.body.removeChild(a);
};

/**
 * Return a formatted isolate name given an ``isolate`` object.
 *
 * @func
 * @param isolate {object}
 * @returns {string}
 */
export const formatIsolateName = isolate => {
    const sourceType = get(isolate, "source_type") || get(isolate, "sourceType");
    const sourceName = get(isolate, "source_name") || get(isolate, "sourceName");

    return sourceType === "unknown" ? "Unnamed" : `${capitalize(sourceType)} ${sourceName}`;
};

/**
 * Transforms a plain taskName (eg. pathoscope_bowtie) to a human-readable name (eg. PathoscopeBowtie).
 *
 * @func
 * @param taskName {string} plain task name
 * @returns {string}
 */
export const getTaskDisplayName = taskName => get(taskDisplayNames, taskName, startCase(taskName));

export const reportAPIError = action => window.Raven.captureException(action.error);

export const routerLocationHasState = (state, key, value) =>
    !!state.router.location.state &&
    (value ? state.router.location.state[key] === value : !!state.router.location.state[key]);

/**
 * Returns an action creator that returns an action with ``type`` as the only property.
 *
 * @func
 * @param type {string} the value to use for the type property
 * @returns {function}
 */
export const simpleActionCreator = type =>
    function actionCreator() {
        return { type };
    };

export const getTargetChange = target => ({
    name: target.name,
    value: target.value,
    error: `error${upperFirst(target.name)}`
});

/**
 * Object that maps algorithm names (task names) to human-readable names.
 *
 * @type {object}
 */
export const taskDisplayNames = {
    aodp: "AODP",
    create_sample: "Create Sample",
    create_subtraction: "Create Subtraction",
    nuvs: "NuVs",
    pathoscope_bowtie: "Pathoscope",
    pathoscope_snap: "Pathoscope",
    build_index: "Build Index"
};

export const toThousand = number => numbro(number).format({ thousandSeparated: true });

/**
 * Converts a ``number`` to a scientific notation string.
 *
 * @func
 * @param {number} number
 * @returns {string}
 */
export const toScientificNotation = number => {
    if (number < 0.01 || number > 1000) {
        const [coefficient, exponent] = split(number.toExponential(), "e");
        return `${numbro(coefficient).format("0.00")}E${replace(exponent, "+", "")}`;
    }

    return numbro(number).format("0.000");
};

export const checkAdminOrPermission = (state, permission) =>
    getAccountAdministrator(state) || state.account.permissions[permission];
