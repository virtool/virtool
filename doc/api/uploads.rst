.. _uploads:

Uploads
=======

When uploads are required in Virtool, they should be targeted at ``/upload``. The available endpoints are:

+-------------------------------------------------+-------------------------+
| Situation                                       | Endpoint                |
+=================================================+=========================+
| Upload and import a Virtool viruses export file | /upload/viruses         |
+-------------------------------------------------+-------------------------+
| Upload a Illumina read file                     | /upload/reads           |
+-------------------------------------------------+-------------------------+
| Upload a .hmm file for use with NuVs            | /upload/hmm/profiles    |
+-------------------------------------------------+-------------------------+
| Upload and import a NuVs annotation file        | /upload/hmm/annotations |
+-------------------------------------------------+-------------------------+
| Upload a host FASTA file                        | /upload/host            |
+-------------------------------------------------+-------------------------+

