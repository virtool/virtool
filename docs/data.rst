Data Layer
==========

.. automodule:: virtool.data.__init__
    :members:

Transforms
----------

.. automodule:: virtool.data.transforms


API
+++

.. autoclass:: virtool.data.transforms.AbstractTransform
    :members:

.. autofunction:: virtool.data.transforms.apply_transforms

Errors
------

.. automodule:: virtool.data.errors
    :members:


