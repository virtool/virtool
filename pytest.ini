[pytest]
addopts = --aiohttp-loop uvloop --capture tee-sys
asyncio_mode = auto
