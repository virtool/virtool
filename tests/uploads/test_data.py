import datetime
from pathlib import Path

import pytest
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncEngine, AsyncSession
from syrupy import SnapshotAssertion
from syrupy.matchers import path_type

from virtool.data.errors import ResourceNotFoundError
from virtool.data.layer import DataLayer
from virtool.fake.next import DataFaker, fake_file_chunker
from virtool.pg.utils import get_row_by_id
from virtool.uploads.models import SQLUpload, UploadType


async def test_create(
    data_path: Path,
    data_layer: DataLayer,
    example_path: Path,
        fake: DataFaker,
    pg: AsyncEngine,
    snapshot,
):
    user = await fake.users.create()

    fake_file_path = example_path / "reads/single.fq.gz"

    upload = await data_layer.uploads.create(
        fake_file_chunker(fake_file_path),
        "sample_1.fq.gz",
        UploadType.reads,
        user=user.id,
    )

    assert upload == snapshot(
        name="obj",
        matcher=path_type(
            {"created_at": (datetime.datetime,), "uploaded_at": (datetime.datetime,)},
        ),
    )

    assert (await get_row_by_id(pg, SQLUpload, upload.id)).to_dict() == snapshot(
        name="sql",
        matcher=path_type(
            {"created_at": (datetime.datetime,), "uploaded_at": (datetime.datetime,)},
        ),
    )

    assert (
        open(data_path / "files" / upload.name_on_disk, "rb").read()
        == open(fake_file_path, "rb").read()
    )


async def test_delete(
    data_path: Path,
    data_layer: DataLayer,
        fake: DataFaker,
    snapshot_recent: SnapshotAssertion,
):
    before = await fake.uploads.create(user=await fake.users.create())

    path = data_path / "files" / before.name_on_disk

    assert path.is_file()

    assert before.removed_at is None

    after = await data_layer.uploads.delete(before.id)

    assert after == snapshot_recent(
        name="after",
    )

    assert not path.is_file()

    with pytest.raises(ResourceNotFoundError):
        await data_layer.uploads.get(before.id)

    with pytest.raises(ResourceNotFoundError):
        await data_layer.uploads.delete(before.id)


@pytest.mark.parametrize("multi", [True, False])
async def test_release(
    multi: bool,
    data_layer: DataLayer,
        fake: DataFaker,
    pg: AsyncEngine,
):
    user = await fake.users.create()

    upload_1 = await fake.uploads.create(user=user, reserved=True)
    upload_2 = await fake.uploads.create(user=user, reserved=True)
    upload_3 = await fake.uploads.create(user=user, reserved=True)

    assert all([upload_1.reserved, upload_2.reserved, upload_3.reserved])

    assert (
        await data_layer.uploads.release(
            [upload_1.id, upload_3.id] if multi else upload_2.id,
        )
        is None
    )

    async with AsyncSession(pg) as session:
        result = await session.execute(select(SQLUpload).order_by(SQLUpload.id))
        reserved = [u.reserved for u in result.unique().scalars().all()]

    assert reserved == [False, True, False] if multi else [True, False, True]
