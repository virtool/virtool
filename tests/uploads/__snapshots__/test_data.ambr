# serializer version: 1
# name: test_create[obj]
  dict({
    'created_at': datetime,
    'id': 1,
    'name': 'sample_1.fq.gz',
    'name_on_disk': '1-sample_1.fq.gz',
    'ready': True,
    'removed': False,
    'removed_at': None,
    'reserved': False,
    'size': 16700094,
    'type': 'reads',
    'uploaded_at': datetime,
    'user': dict({
      'handle': 'leeashley',
      'id': 'bf1b993c',
    }),
  })
# ---
# name: test_create[sql]
  dict({
    'created_at': datetime,
    'id': 1,
    'name': 'sample_1.fq.gz',
    'name_on_disk': '1-sample_1.fq.gz',
    'ready': True,
    'removed': False,
    'removed_at': None,
    'reserved': False,
    'size': 16700094,
    'type': 'reads',
    'uploaded_at': datetime,
    'user': 'bf1b993c',
  })
# ---
# name: test_delete[after]
  dict({
    'created_at': datetime,
    'id': 1,
    'name': 'test.fq.gz',
    'name_on_disk': '1-test.fq.gz',
    'ready': True,
    'removed': True,
    'removed_at': datetime,
    'reserved': False,
    'size': 16700094,
    'type': 'reads',
    'uploaded_at': datetime,
    'user': dict({
      'handle': 'leeashley',
      'id': 'bf1b993c',
    }),
  })
# ---
